Support NVR Hikvision devices

ptarjan opened the pull request:
This commit adds:
- NVR helper functions for extended event detection with broader notification method support (record, email, beep in addition to center and HTTP)
- Camera platform support for Hikvision NVR and cameras with RTSP streaming and HTTP snapshot capabilities
- Video channel discovery for NVRs with multiple cameras
- Device class mapping for Hikvision event types

The NVR event detection extends pyhik's capabilities by accepting additional notification methods commonly used on NVRs but ignored by the pyhik library.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/42614
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [ ] I understand the code I am submitting and can explain how it works.
- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.
- [ ] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

joostlek commented on homeassistant/components/hikvision/__init__.py:
```diff
@@ -53,21 +59,36 @@ async def async_setup_entry(hass: HomeAssistant, entry: HikvisionConfigEntry) ->
             HikCamera, url, port, username, password
         )
-    except requests.exceptions.RequestException as err:
+    except Exception as err:
```
We should be specific here and not catch bare exceptions

joostlek commented on homeassistant/components/hikvision/binary_sensor.py:
```diff
@@ -177,6 +138,7 @@ def __init__(
         channel: int,
     ) -> None:
         """Initialize the binary sensor."""
+        self._entry = entry
```
idem

joostlek commented on homeassistant/components/hikvision/camera.py:
```diff
@@ -0,0 +1,183 @@
+            self._attr_name = channel.name
+        else:
+            # Single camera - use None to just use device name
+            self._attr_name = None
```
What is a channel exactly? Is a channel a separate camera in a separate room? Should they be their own device?

joostlek commented on homeassistant/components/hikvision/camera.py:
```diff
@@ -0,0 +1,183 @@
+                "Error getting snapshot from %s: %s", self._channel.name, err
+            )
+
+        return self._last_image
```
Should this raise HomeAssistantError instead?

joostlek commented on homeassistant/components/hikvision/camera.py:
```diff
@@ -0,0 +1,183 @@
+        return {
+            "channel_id": self._channel.id,
+            "channel_name": self._channel.name,
+        }
```
What's the benefit of a channel id? The name is the entiy name

ptarjan committed:
Use pyhik built-in methods for NVR event handling

- Upgrade pyhik to 0.3.4 which includes built-in methods for NVR event handling
- Use get_event_triggers() and inject_events() from pyhik for NVR devices
- Fix test translation issues with proper ignore_missing_translations
- Add comprehensive tests for NVR event fetching

ptarjan commented:
Sorry for the noise in here :(. I did this before I pulled all the helper code into the included library. It should be so much simpler now.

joostlek commented on homeassistant/components/hikvision/manifest.json:
```diff
@@ -8,5 +8,5 @@
   "loggers": ["pyhik"],
   "quality_scale": "legacy",
-  "requirements": ["pyHik==0.3.2"]
+  "requirements": ["pyHik==0.3.4"]
```
Can we bump in a separate PR? Can you also create a separate PR where you add yourself as codeowner as you are now helping a lot with the integration?

joostlek commented on tests/components/hikvision/conftest.py:
```diff
@@ -61,7 +61,6 @@ def mock_hikcamera() -> Generator[MagicMock]:
     with (
         patch(
             "homeassistant.components.hikvision.HikCamera",
-            autospec=True,
```
why do we remove the autospec? It generally preshapes your mock so you don't have to prepare everything

joostlek commented on tests/components/hikvision/test_init.py:
```diff
@@ -89,3 +89,19 @@ async def test_setup_entry_no_device_id(
+) -> None:
+    """Test setup fetches NVR events for NVR devices."""
+    mock_hikcamera.return_value.get_type = "NVR"
+    mock_hikcamera.return_value.get_event_triggers.return_value = {"Motion": [1, 2]}
```
How often will you test with the NVR? You could also extend the `mock_hikcamera` and create a `mock_hik_nvr` that imports `mock_hikcamera` and sets the right fields for the NVR. This way you don't have to repeat yourself

joostlek reviewed (changes requested):
The title is not updated

ptarjan committed:
Address code review feedback

- Group executor jobs for NVR event fetching to reduce event loop switching
- Restore autospec=True on mock_hikcamera fixture
- Add mock_hik_nvr fixture for NVR-specific tests
